### DISCUSSION TITLE
Request for Inclusion of Horizontal Pod Autoscaling (HPA) Option in KubernetesPodOperator

### DISCUSSION THREAD

**Discussion by codeprasan** [apache/airflow#36939] [General]
### Description

In the current Airflow version, the KubernetesPodOperator allows users to specify computer resources, but these resources are distributed across nodes. Unfortunately, there is a shortage of available nodes, leading to resource constraints. It would be advantageous to introduce an option for Horizontal Pod Autoscaling (HPA) for this operator. Currently, Airflow, through the KubernetesPodOperator, generates a new pod dynamically for each execution. However, the implementation of HPA/Replicas is hindered by a missing feature or property in the existing operator. The addition of an HPA option would address this limitation and enhance the scalability of the operator.

### Use case/motivation

In our Airflow setup, we aim to trigger Spark jobs, but the absence of Horizontal Pod Autoscaling (HPA) functionality in the KubernetesPodOperator poses a challenge. The current operator lacks the capability to leverage HPA for dynamic scaling, hindering our ability to efficiently scale resources based on workload demands. This limitation impacts the scalability and performance optimization of Spark jobs within the Airflow environment. Introducing HPA support to the KubernetesPodOperator would be instrumental in addressing this issue and enhancing the overall scalability and flexibility of our Spark job execution.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by raphaelauv** [comment]
the KPO start a pod , that's all and will always

use the `KubernetesCreateResourceOperator` to start something like a `RayJob` or `SparkApplication`

**Reply by potiuk** [reply]
Yep. Converting into discussion in case more is needed, but yes. KPO is for pods not for other resource types.

Ket us no2 @codeprasan if using other options will work for you,

**Comment by codeprasan** [comment]
Thank you, @raphaelauv and @potiuk.

Certainly, we'll explore the usage of KubernetesCreateResourceOperator. I'm curious whether this operator supports Horizontal Pod Autoscaling (HPA)?

As we delve into using KubernetesCreateResourceOperator, could you provide information on whether there are examples or documentation available for this operator? Any guidance or references would be greatly appreciated.

While the KubernetesPodOperator dynamically creates pods, I'm interested in functionality that allows for Kubernetes replicas/HPA, enabling the multiplication of pod instances based on workload demands.

**Reply by potiuk** [reply]
Looks like there eExample DAGS: https://github.com/apache/airflow/tree/providers-cncf-kubernetes/7.13.0/tests/system/providers/cncf/kubernetes 

That's all what we have (on top od docstring).

I think it allows to create any custom resource, so anything should be possible, but i think you need to experiment with it to see if it fits your case. Let us know. And maybe that's a good opportunity as well to contribute back better documentation / guidelines anout this case - Airflow (including docs) is created by almost 2800 contributors, and many of them were users who contributed nice documentation back after exploring and finding out how things work in "real" use cases. 

Any nice documentation about that case is welcome.
